### PR TITLE
Query optimization - Remove SELECT CASE for boolean values in WHERE clause

### PR DESCRIPTION
**Pull request by mikary**
Fix for #4304
@anpete @divega

### CODE REVIEW

**Comment by mikary** [comment]
Could you also take a look @maumar

**Comment by maumar** [comment]
:shipit:

File: src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs [hunk 1]
```diff
@@ -1069,6 +1069,15 @@ protected override Expression VisitUnary(UnaryExpression expression)
                         return VisitIsNotNull(isNullExpression);
                     }
 
+                    if(expression.Operand is ExistsExpression)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/Sql/DefaultQuerySqlGenerator.cs]: Formatting.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/ConditionalRemovingExpressionVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+{
+    public class ConditionalRemovingExpressionVisitor : ExpressionVisitorBase
+    {
+        public override Expression Visit([CanBeNull] Expression node)
+        {
+            var selectExpression = node as SelectExpression;
+
+            var conditionalExpression = selectExpression?.Projection.FirstOrDefault() as ConditionalExpression;
+
+            if (selectExpression?.Projection.Count == 1
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/ConditionalRemovingExpressionVisitor.cs]: Duplicate null check of selecExpression. Suggest doing the projection count test before pulling out the conditional expression.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/ConditionalRemovingExpressionVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                && conditionalExpression != null
+                && conditionalExpression.Type == typeof(bool)
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/Internal/ConditionalRemovingExpressionVisitor.cs]: Could use null-conditional here and omit the null comparison above.

**Comment by anpete** [comment]
:shipit:

**Pull request merged**